tdms/tests/system/test_regen.py

```diff
@@ -0,0 +1,112 @@
+    # Fetch test run information
+    with open(config_file_path, "r") as opened_config_file:
+        # Read data into dictionary, and take the "tests" value
+        INFO = yaml.safe_load(opened_config_file)
+        yaml_test_id = INFO["test_id"]
+        if yaml_test_id != test_id:
+            warn(
+                f"{str(config_file_path)} implicit ID does not match config file description (found {test_id})"
+            )
+        run_information = INFO["tests"]
+
+    # Regenerate the input data for arc_{test_id}
+    # Error on not successful (hence fail test)
+    regenerate_test(config_file_path)
+
+    # Perform each run of TDMS as specified by the config file
+    system_test = TDMSSystemTest(yaml_test_id, run_information)
+
+    # Run all of the system tests
+    run_success = system_test.perform_all_runs()
+
+    # TEAR-DOWN: remove the regenerated inputs and outputs from the data/input_generation/arc_{test_id} folder.
+    # To be safe, we can just remove all .mat files from this directory the the subdirectories, since these should be the only system-test TDMS artefacts
+    if not PRESERVE_FLAG:
+        input_data_dump = str(path_to_input_generation / f"arc_{test_id}/*.mat")
+        mat_artefacts = glob(input_data_dump, recursive=True)
+        for mat_file in mat_artefacts:
+            os.remove(mat_file)
+
+    # Although we should have check-ed whether each run was a pass/fail, we can also assert that all runs need to pass here to report failures
+    failed_run_names = []
+    for i, success in enumerate(run_success):
+        if not success:
+            failed_run_names.append(system_test.tdms_runs[i].tdms_run.run_id)
+    assert all(
+        run_success
+    ), f"arc_{test_id} : Some runs were unsuccessful ({len(run_success)-sum(run_success)}/{len(run_success)}) :\n {failed_run_names}"
+    return
+
+
+# Run each system test that is currently saved in the repository, as identified through the config files
+# "MATLAB doesn't run on GH runners. If you want to check that regenerating the input data still allows the tests to pass, run this locally and remove the skip mark."
+@pytest.mark.parametrize("test_id", TEST_IDS)
+def test_system(test_id) -> None:
+    """Runs the system test arc_{test_id}, including fetching missing reference data from Zenodo.
+
+    Wraps the workflow() method, which actually does the bulk of the testing.
+    """
+    print(f"\nRunning {test_id}", end=" | ")
+
+    # the reference OUTPUT data should be at this location
+    ZIP_PATH = ZIP_DESTINATION / f"arc_{test_id}.zip"
+    # download data if not present
+    if not ZIP_PATH.exists():
+        url = ZENODO_URL + f"/arc_{test_id}.zip"
+        print(f"Downloading from {url}")
+        download_data(url, to=ZIP_PATH)
+    else:
+        print(f"Using cache in {ZIP_PATH}")
+
+    # Run the test workflow, for this test
+    workflow(test_id)
+    # End of system test
+    return
```

Comment:
I anticipated that `test_system` and `test_regen` will diverge very quickly as more tests (requiring different input methods) are added, which they do: final version of `test_regen` [is here](https://github.com/UCL/TDMS/blob/wgraham-input_data_fn_docs/tdms/tests/system/test_regen.py). I'd argue that the workflows are sufficiently divergent as to warrant two separate files, but there are some other supporting reasons:
- `test_regen`'s workflow is already different to `test_system`'s. Additionally, the `pytest` parametrisation also changes to use the `config` files rather than relying on the manually-typed `Zenodo` list
- `workflow` exists to provide some distinction between the fetch-request for the data and the actual steps in the test. Granted, not the most informative name, but I would want to separate it from the setup steps regardless.
- `test_system.py` (and it's supporting classes) will be made obsolete if we ever convert the system tests entirely into Python or get MATLAB to behave on runners. In which case, I'd rather have `test_regen` working stand-alone so we don't get surprised when other files get deleted. (Also applies to a couple of the other classes in the files below)